### ISSUE TITLE
Last rebase from July 6 introduces a ~10% memory regression

### ISSUE THREAD

**Issue by rafaelauler** [facebookincubator/BOLT#188]
For some workloads. We know at the moment that this is coming from changes in LLVM from May 25th to July 6th.

**Comment by rafaelauler** [comment]
Blame: https://reviews.llvm.org/D76877

After reverting this diff, I was able to avoid the regression observed in some workloads affected by it.

**Comment by h-vetinari** [comment]
> Blame: https://reviews.llvm.org/D76877

Have you raised an issue upstream for this?

**Comment by rafaelauler** [comment]
Hi @h-vetinari, thanks for asking, let me write the resolution for this investigation and then I'll close this.

I opened this investigation because we had 3 memory regressions at the same time, and this was one of them. Regression 1 came from a problem in our test builds that accidentally started building large executables with -O0, generating binaries with way more functions. Our monitoring systems detected this as a regression (BOLT memory utilization increases with the number of functions in the binary), but most of the memory increase wasn't caused by BOLT.
Regression 2, this one here, *only* manifested in these non-standard -O0 builds and is the part of the regression that was caused by BOLT. After investigating it, it was blamed to that diff. The reason that this diff is important is because a significant part of the memory footprint was being used by FDE records, since the binary had many functions, but all of them were cold, and the bulk of the metadata that we were maintaining for them were FDE records. These records increased in size after this upstream diff. However, in the end I wasn't worried about it, because we never run BOLT on -O0 builds, so it became less important to protest this diff (and I particularly don't have any good ideas on how to implement its functionality without increasing memory footprint a bit).
Regression 3 was caused only for binaries that use debug fission (split dwarf). This was diagnosed and fixed.
